src/vm_pages.rs

```diff
@@ -1264,6 +1319,75 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
         self.do_remove_region(page_addr, len, VmRegionType::Mmio)
     }
 
+    /// Converts the specified memory region from confidential to shared. Returns the TLB version
+    /// at which the conversion will be completed.
+    pub fn share_mem_region(&self, page_addr: GuestPageAddr, len: u64) -> Result<TlbVersion> {
+        let end = PageAddr::new(
+            RawAddr::from(page_addr)
+                .checked_increment(len)
+                .ok_or(Error::AddressOverflow)?,
+        )
+        .ok_or(Error::UnalignedAddress)?;
+        // Check the address range lies within a "Confidential" region of guest physical address
+        // space and update the region type to "ConfidentialRemovable".
+        self.inner.regions.write().update_in_place(
+            page_addr,
+            end,
+            VmRegionType::Confidential,
+            VmRegionType::ConfidentialRemovable,
+            false,
+        )?;
+
+        // Should we go further here, by getting the actual list of pages that will be
+        // invalidated, so that we can know for sure if we can return to the guest.
+        Ok(self.inner.tlb_tracker.current_version().increment())
+    }
+
+    /// Completes guest memory sharing by verifying the pages have been previously removed
+    /// and by marking the region as "Shared".
+    pub fn complete_share_mem_region(&self, page_addr: GuestPageAddr, len: u64) -> Result<()> {
+        let end = PageAddr::new(
+            RawAddr::from(page_addr)
+                .checked_increment(len)
+                .ok_or(Error::AddressOverflow)?,
+        )
+        .ok_or(Error::UnalignedAddress)?;
+        // Check the address range lies within a "ConfidentialRemovable" region of guest
+        // physical address space and update the region type to "Shared".
+        self.inner.regions.write().update_in_place(
+            page_addr,
+            end,
+            VmRegionType::ConfidentialRemovable,
+            VmRegionType::Shared,
+            true,
+        )?;
```

Comment:
need to check that all the pages have been removed first

Reply:
Ah yes of course, given we've cleared the PTE bits through `remove_pages`, we expect every PTE from the range to be marked as `Unused`.